### DISCUSSION TITLE
Better way to create objects

### DISCUSSION THREAD

**Discussion by Brosbud** [Pi4J/pi4j#219] [Q&A] [answered]
Hi! 
I just recently got the library working, and have been playing around with it for a little bit, and i have 2 questions,

1, What exactly do the parametres do, like i know .address tells what pin it is, 
but what does .id or .name do?

2. Can the object creation be externalized to for example another class? When i only have to create one object, its fine, but sometimes i need to make alot of objects, and the most compact i can get it is;

 var ledAconf = DigitalOutput.newConfigBuilder(pi4j)
                .address(A)
                .shutdown(DigitalState.LOW)
                .provider("pigpio-digital-output");
var ledA = pi4j.create(ledAconf);

And that really stacks up

Edit; And is it possible to create the objects as "universal", meaning they could be used outside the mainclass, and in methods

Thanks in advance!

**Comment by savageautomate** [comment, marked as answer]
@Brosbud 

One of the Pi4J V2 goals was to have default providers based on the current "platform".  However, this is not currently implemented.  This would avoid having to explicitly specify which provider to use and could result in GPIO instance creation as simple as:

```
        var pin0 = pi4j.dout().create(0);
        var pin1 = pi4j.dout().create(1);
        var pin2 = pi4j.dout().create(2);
        var pin3 = pi4j.dout().create(3);

```

We are working on this "default" behaviour now.

Another option is to reuse the `config` object that you create overriding the 'address' (and 'id' if used) for each I/O instance:

```
        var config = DigitalOutput.newConfigBuilder(pi4j)
                .provider("linuxfs-digital-output")
                .shutdown(DigitalState.LOW)
                .initial(DigitalState.LOW);

        var pin0 = pi4j.dout().create(config.address(0).id("my-led"));
        var pin1 = pi4j.dout().create(config.address(1).id("my-relay"));
        var pin2 = pi4j.dout().create(config.address(2).id("my-lock"));
        var pin3 = pi4j.dout().create(config.address(3).id("my-pump"));

```

You could also create I/O instances via a set of "Properties" like in this examples code:
https://github.com/Pi4J/pi4j-example-telegraph/blob/master/src/main/java/com/pi4j/demo/telegraph/TelegraphUsingProperties.java

**Reply by Brosbud** [reply]
Is there a special purpose to the .id tag?

**Comment by savageautomate** [comment]
@Brosbud 

Yes. the `id` field is used internally inside the Pi4J context/runtime to keep track of the instances.  If you don't assign an `id`, Pi4J will create a unique ID string for the instance -- so its optional if you want to specify your own unique ID string.  Additionally, you can retrieve the I/O instance from the Pi4J context  anywhere else in your program if needed by the `id`.  This can help in some cases where you only need to pass around the single Pi4J context and you can still gain access to the I/O instances without having to track and pass around your own variable references.  So even if your variable reference to an I/O instance goes out of scope, Pi4J will maintain a reference to the I/O instance internally until it is `shutdown()`.

```java
// three ways to get existing LED output instance
DigitalOutput led = pi4j.io("my-led");
var led = (DigitalOutput)pi4j.io("my-led");
var led = pi4j.io("my-led", DigitalOutput.class);
```

Internally the I/O instances are maintained by the `Registry`.   You can gain access to the `Registry` via the context.  There are additional methods in registry to interrogate/discover/enumerate the created I/O instances at runtime.

```java
// get the Pi4J I/O registry
Registry registry = pi4j.registry();

// check to see if the LED output already exists (by id)
boolean myLedAlreadyExists = registry.exists("my-led");

// get all digital output instances from the Pi4J I/O registry       
var outputs = registry.allByIoType(IOType.DIGITAL_OUTPUT);
``` 

Fields like `name` and `description` are entirely optional and not used by Pi4J internally except to print if performing a `describe()` or `toString()` operation on Pi4J objects.


```java
// create digital output I/O configuration
var config = DigitalOutput.newConfigBuilder(pi4j)
                        .id("my-dout")
                        .name("My Digital Output")
                        .address(GPIO_PIN)
                        .shutdown(DigitalState.LOW)
                        .initial(DigitalState.HIGH)
                        .provider("linuxfs-digital-output");

// create digital output I/O instance using configuration
var output = pi4j.create(config);

// print digital output object to system out
output.describe().print(System.out);

----

// ... CONSOLE OUTPUT
// > IO: "My Digital Output" {my-dout} <com.pi4j.plugin.linuxfs.provider.gpio.digital.LinuxFsDigitalOutput> {DOUT-26} 

```

**Comment by FDelporte** [comment]
Thanks @Brosbud for the question and @savageautomate for the clear answers. This was exactly the info that was still missing on our website as there was only "TODO" on this page... https://pi4j.com/documentation/build-io/

I used the info from this discussion to fill that page... ;-)
